dumang_ctrl/dumang/common.py

```diff
@@ -481,22 +536,29 @@ def configured_key(self, k):
     @property
     def configured_keys(self):
         if not self._keys_initialized:
+            pending = 0
             for k in range(MAX_KEYS):
                 self.put(KeyReportRequestPacket(k))
-
-            while True:
-                # NOTE: Wait for all requests to be sent out and the recv queue to be populated.
-                if self.send_q.empty():
-                    while not self.recv_q.empty():
-                        p = self.recv_q.get()
-                        if isinstance(p, KeyReportResponsePacket):
-                            for l, kc in p.layer_keycodes.items():
-                                if kc.keystr is not UNKNOWN_KEYCODE_STR:
-                                    # NOTE: We add the layer_keycodes to the DKM
-                                    p.key.layer_keycodes = p.layer_keycodes
-                                    self._configured_keys[p.key.key] = DuMangKeyModule(p.key, p.layer_keycodes)
-                    self._keys_initialized = True
-                    break
+                pending += 1
+
+            while pending > 0:
+                p = self.recv_q.get()
+                if isinstance(p, KeyReportResponsePacket):
+                    pending -= 1
+                    if any([kc.keycode != 0 for kc in p.layer_keycodes.values()]):
+                        # NOTE: We add the layer_keycodes to the DKM
+                        p.key.layer_keycodes = p.layer_keycodes
+                        self._configured_keys[p.key.key] = DuMangKeyModule(p.key, p.layer_keycodes)
+                    if any([kc.keycode == Keycode.MACRO for kc in p.layer_keycodes.values()]):
+                        self.put(MacroReportRequestPacket(p.key, 0))
+                        pending += 1
+                if isinstance(p, MacroReportResponsePacket):
+                    pending -= 1
+                    if p.type.type not in [0, 0xff]:
```

Comment:
Which `type` is this referring to? Just trying to get some understanding on the `0` to `0xff`.

Reply:
it's MacroType.type.
while i don't really know what 0 and 0xff are,
they seemed like a termination signal of a series of packets.